src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,703 @@
+  dfcc_cfg_infot &cfg_info,
+  const std::set<symbol_exprt> &local_statics,
+  std::set<irep_idt> &function_pointer_contracts)
+{
+  const dfcc_loop_infot &loop = cfg_info.get_loop_info(loop_id);
+  const std::size_t cbmc_loop_id = loop.cbmc_loop_id;
+  const exprt &outer_write_set = cfg_info.get_outer_write_set(loop_id);
+  const irep_idt language_mode = utils.get_function_symbol(function_id).mode;
+  goto_programt::targett head = loop.find_head(goto_function.body).value();
+  auto head_location(head->source_location());
+
+  auto &symbol_table = goto_model.symbol_table;
+
+  // Temporary variables:
+  // Create a temporary to track if we entered the loop,
+  // i.e., the loop guard was satisfied.
+  const auto entered_loop =
+    new_tmp_symbol(
+      bool_typet(),
+      head_location,
+      language_mode,
+      symbol_table,
+      std::string(ENTERED_LOOP) + "__" + std::to_string(cbmc_loop_id))
+      .symbol_expr();
+
+  // Create a snapshot of the invariant so that we can check the base case,
+  // if the loop is not vacuous and must be abstracted with contracts.
+  const auto initial_invariant =
+    new_tmp_symbol(
+      bool_typet(), head_location, language_mode, symbol_table, INIT_INVARIANT)
+      .symbol_expr();
+
+  // Create a temporary variable to track base case vs inductive case
+  // instrumentation of the loop.
+  const auto in_base_case = new_tmp_symbol(
+                              bool_typet(),
+                              head_location,
+                              language_mode,
+                              symbol_table,
+                              "__in_base_case")
+                              .symbol_expr();
+
+  // Temporary variables for storing the multidimensional decreases clause
+  // at the start of and end of a loop body.
+  exprt::operandst decreases_clauses = loop.decreases;
+  std::vector<symbol_exprt> old_decreases_vars, new_decreases_vars;
+  for(const auto &clause : decreases_clauses)
+  {
+    // old
+    const auto old_decreases_var =
+      new_tmp_symbol(clause.type(), head_location, language_mode, symbol_table)
+        .symbol_expr();
+    old_decreases_vars.push_back(old_decreases_var);
+    // new
+    const auto new_decreases_var =
+      new_tmp_symbol(clause.type(), head_location, language_mode, symbol_table)
+        .symbol_expr();
+    new_decreases_vars.push_back(new_decreases_var);
+  }
+
+  // Convert the assigns clause to the required type.
+  exprt::operandst assigns(loop.assigns.begin(), loop.assigns.end());
+
+  // Add local statics to the assigns clause.
+  for(auto &local_static : local_statics)
+  {
+    assigns.push_back(local_static);
+  }
+
+  auto nof_targets = assigns.size();
+  max_assigns_clause_size = std::max(nof_targets, max_assigns_clause_size);
+
+  //   __populate(w_loop, <loop_assigns>);
+  // Populating the write set from loop assigns target.
+  goto_programt assigns_instrs;
+  goto_programt write_set_populate_instrs;
+
+  //   __havoc(w_loop);
```

Comment:
Similar to the above: I don't understand what `__havoc` means.

Reply:
I modified it to just `havoc(w_loop)`. The instructions havoc the write set of the loop.